src/hoisted.jl

```diff
@@ -0,0 +1,84 @@
+# Specialized functions for increased performance when JSON is in-memory
+
+addbyte!(x::Int, c) = x + 1
+addbyte!(x::Vector{UInt8}, c) = push!(x, c)
+
+function parse_string(ps::MemoryParserState)
+    # "Dry Run": find length of string so we can allocate the right amount of
+    # memory from the start.
+    s = ps.s
+    len = parse_string(ps, 0)
+
+    # Now read the string itself
+    ps.s = s
+    ret = Compat.UTF8String(UInt8[])
+    b = sizehint!(ret.data, len)
+    parse_string(ps, b)
+    ret
```

Comment:
Seems `:jl_array_to_string` makes the intent more clear, so I'll switch to that anyway. Agreed on the `unsafe_array_to_string`.